dialer_test.go

```diff
@@ -0,0 +1,155 @@
+package cloudsqlconn
+
+import (
+	"context"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"testing"
+
+	"cloud.google.com/cloudsqlconn/internal/mock"
+	"google.golang.org/api/option"
+	sqladmin "google.golang.org/api/sqladmin/v1beta4"
+)
+
+// replaceSQLAdminClient patches the Dialer's SQL Admin API client with one that
+// uses the provided HTTP client and endpoint.
+func replaceSQLAdminService(d *Dialer, client *http.Client, endpoint string) {
+	svc, err := sqladmin.NewService(
+		context.Background(),
+		option.WithHTTPClient(client),
+		option.WithEndpoint(endpoint),
+	)
+	if err != nil {
+		panic(err)
+	}
+	d.sqladmin = svc
+}
+
+func TestDialerCanConnectToInstance(t *testing.T) {
+	inst := mock.NewFakeCSQLInstance("my-project", "my-region", "my-instance")
+	mc, url, cleanup := mock.HTTPClient(
+		mock.InstanceGetSuccess(inst, 1),
+		mock.CreateEphemeralSuccess(inst, 1),
+	)
+	stop := mock.StartServerProxy(mock.ServerProxyConfig{
+		Response: "server-response",
+		Instance: inst,
+	})
+	defer func() {
+		stop()
+		if err := cleanup(); err != nil {
+			t.Fatalf("%v", err)
+		}
+	}()
+
+	d, err := NewDialer(context.Background(), WithDefaultDialOptions(WithPublicIP()))
+	if err != nil {
+		t.Fatalf("expected NewDialer to succeed, but got error: %v", err)
+	}
+	replaceSQLAdminService(d, mc, url)
+
+	conn, err := d.Dial(context.Background(), "my-project:my-region:my-instance", WithPublicIP())
+	if err != nil {
+		t.Fatalf("expected Dial to succeed, but got error: %v", err)
+	}
+	defer conn.Close()
+
+	data, err := ioutil.ReadAll(conn)
+	if err != nil {
+		t.Fatalf("expected ReadAll to succeed, got error %v", err)
+	}
+	if string(data) != "server-response" {
+		t.Fatalf("expected known response from the server, but got %v", string(data))
+	}
+}
+
+func TestDialerInstantiationErrors(t *testing.T) {
+	_, err := NewDialer(context.Background(), WithCredentialsFile("bogus-file.json"))
+	if err == nil {
+		t.Fatalf("expected NewDialer to return error, but got none.")
+	}
+}
+
+func errorContains(err error, want string) bool {
+	return strings.Contains(err.Error(), want)
+}
+
+func TestDialWithAdminAPIErrors(t *testing.T) {
+	inst := mock.NewFakeCSQLInstance("my-project", "my-region", "my-instance")
+	mc, url, cleanup := mock.HTTPClient()
+	stop := mock.StartServerProxy(mock.ServerProxyConfig{
+		Instance: inst,
+	})
+	defer func() {
+		stop()
+		_ = cleanup() // skip mock verification of HTTP methods
```

Comment:
We shouldn't skip this here, because if it throws an error it means an unexpected call was made.

Reply:
Turns out, I had mistakenly registered mocks in a previous attempt. I'll add the check back.